core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowCompiler.scala

```diff
@@ -15,21 +15,105 @@ import edu.uci.ics.texera.web.workflowruntimestate.WorkflowAggregatedState.FAILE
 
+  /**
+    * Compile the workflow to logical plan and errors(if any)
+    * @param logicalPlanPojo the pojo parsed from workflow str provided by user
+    * @return LogicalPlan, and a Map from OpId to Op's error(this map is empty if there is no error)
+    */
+  def compileToLogicalPlan(
+      logicalPlanPojo: LogicalPlanPojo
+  ): (LogicalPlan, Map[OperatorIdentity, WorkflowFatalError]) = {
+    val errorList = new ArrayBuffer[(OperatorIdentity, Throwable)]()
+    val opIdToError = mutable.Map[OperatorIdentity, WorkflowFatalError]()
+
+    var logicalPlan: LogicalPlan = LogicalPlan(logicalPlanPojo)
+    logicalPlan = SinkInjectionTransformer.transform(
+      logicalPlanPojo.opsToViewResult,
+      logicalPlan
+    )
+
+    logicalPlan.propagateWorkflowSchema(context, Some(errorList))
+    // report compilation errors
+    if (errorList.nonEmpty) {
+      errorList.foreach {
+        case (opId, err) =>
+          logger.error("error occurred in logical plan compilation", err)
+          opIdToError += (opId -> WorkflowFatalError(
+            COMPILATION_ERROR,
+            Timestamp(Instant.now),
+            err.toString,
+            getStackTraceWithAllCauses(err),
+            opId.id
+          ))
+      }
+    }
+    (logicalPlan, opIdToError.toMap)
+  }
+
+  /**
+    * Compile a workflow to physical plan, along with the schema propagation result and error(if any)
+    *
+    * @param logicalPlanPojo the pojo parsed from workflow str provided by user
+    * @return WorkflowCompilationResult, containing the physical plan, input schemas per op and error per op
+    */
+  def compileToPhysicalPlan(
+                             logicalPlanPojo: LogicalPlanPojo
+                           ): WorkflowCompilationResult = {
+    val (logicalPlan, opIdToError) = compileToLogicalPlan(logicalPlanPojo)
+    if (opIdToError.nonEmpty) {
+      // encounter errors during compile pojo to logical plan,
+      //   so directly return empty physical plan, schema map and non-empty error map
+      return WorkflowCompilationResult(
+        physicalPlan = PhysicalPlan.empty,
+        operatorIdToInputSchemas = Map.empty,
+        operatorIdToError = opIdToError
+      )
+    }
+    // from logical plan to physical plan
+    val physicalPlan = PhysicalPlan(context, logicalPlan)
+
+    // Extract physical input schemas, excluding internal ports
+    val physicalInputSchemas = physicalPlan.operators.map { physicalOp =>
+      physicalOp.id -> physicalOp.inputPorts.values
+        .filterNot(_._1.id.internal)
+        .map {
+          case (port, _, schema) => port.id -> schema.toOption
+        }
+    }
+
+    // Group the physical input schemas by their logical operator ID and consolidate the schemas
+    val opIdToInputSchemas = physicalInputSchemas
+      .groupBy(_._1.logicalOpId)
+      .view
+      .mapValues(_.flatMap(_._2).toList.sortBy(_._1.id).map(_._2))
+      .toMap
+
+    WorkflowCompilationResult(physicalPlan, opIdToInputSchemas, Map.empty)
```

Comment:
it is usually a better idea to define a success result and a failure message, this way the json content is more clean and the code is more readable.

For example, it is hard for me as a reviewer or a maintainer to understand if this line is a success return or a failure return, I'd have to understand the payload's semantic in order to judge.

Reply:
How about making the physical plan `Option[PhysicalPlan]`, use `None` or not to indicate the success or failure?